File: backend/src/models/primary/class.rs
use crate::helpers::null_to_default;
use crate::models::{
    Attribute, LinkPreview, Proficiency, Publication, Rule, Skill, Summary, SummaryData,
};
use crate::storeable::Storeable;
use bson::oid::ObjectId;
use serde::{Deserialize, Serialize};
use specta::Type;

use super::Feat;

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Class {
    #[serde(rename = "_id")]
    pub mongo_id: ObjectId,
    pub fvtt_id: String,
    pub name: String,
    #[serde(default, deserialize_with = "null_to_default")]
    pub ancestry_feat_levels: Vec<u8>,
    pub attacks: ClassAttacks,
    #[serde(default, deserialize_with = "null_to_default")]
    pub class_feat_levels: Vec<u8>,
    pub defenses: ClassDefenses,
    pub description: String,
    #[serde(default, deserialize_with = "null_to_default")]
    pub general_feat_levels: Vec<u8>,
    pub hp: u8,
    #[serde(default, deserialize_with = "null_to_default")]
    pub features: Vec<Feat>,
    #[serde(default, deserialize_with = "null_to_default")]
    pub key_ability: Vec<Attribute>,
    pub perception: Proficiency,
    pub publication: Publication,
    #[serde(default, deserialize_with = "null_to_default")]
    pub rules: Vec<Rule>,
    pub saving_throws: ClassSavingThrows,
    #[serde(default, deserialize_with = "null_to_default")]
    pub skill_feat_levels: Vec<u8>,
    #[serde(default, deserialize_with = "null_to_default")]
    pub skill_increase_levels: Vec<u8>,
    pub spellcasting: bool,
    pub trained_skills: ClassTrainedSkills,
    pub rarity: String,
    #[serde(default, deserialize_with = "null_to_default")]
    pub traits: Vec<String>,
    pub slug: String,
    #[serde(rename = "type")]
    pub data_type: String,
}

impl Storeable for Class {
    fn table_name() -> &'static str {
        "class"
    }
}

impl From<Class> for Summary {
    fn from(value: Class) -> Self {
        let data = vec![SummaryData::String {
            value: value.name.clone(),
            link: None,
            tooltip: None,
            abbreviation: None,
        }];

        Self {
            id: value.fvtt_id,
            name: value.name,
            slug: value.slug,
            data,
        }
    }
}

impl From<Class> for LinkPreview {
    fn from(value: Class) -> Self {
        Self {
            slug: value.slug,
            name: value.name,
            description: value.description,
            rarity: Some(value.rarity),
            traits: value.traits,
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ClassAttacks {
    pub unarmed: bool,
    pub simple: bool,
    pub martial: bool,
    pub advanced: bool,
    pub other: ClassAttacksOther,
}

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ClassAttacksOther {
    pub name: String,
    pub rank: Proficiency,
}

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ClassDefenses {
    pub unarmored: bool,
    pub light: bool,
    pub medium: bool,
    pub heavy: bool,
}

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ClassSavingThrows {
    pub fortitude: Proficiency,
    pub reflex: Proficiency,
    pub will: Proficiency,
}

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ClassTrainedSkills {
    pub additional: u8,
    pub value: Vec<Skill>,
    pub custom: Option<String>,
}

File: data-transformer/src/models/class.ts
import { Exclude, Expose, Transform, Type } from "class-transformer";
import { assignHeaderIds, cleanupHTML, stepHeaders } from "./utils/html";
import type { Publication } from "./secondary/publication";
import { Feat } from "./feat";
import { CLASS_FEATURES } from ".";
import { extractLastUuid } from "./utils/uuid";
import type { Item } from "./secondary/item";
import { mapToRule } from "./secondary/rule";

export class Class {
    @Expose({ name: "_id" }) fvttId!: string;
    name!: string;

    @Transform(({ obj }) => obj.system.ancestryFeatLevels.value)
    @Expose()
    ancestryFeatLevels!: number[];

    @Transform(({ obj }) => {
        return {
            unarmed: obj.system.attacks.unarmed,
            simple: obj.system.attacks.simple,
            martial: obj.system.attacks.martial,
            advanced: obj.system.attacks.advanced,
            other: obj.system.attacks.other,
        };
    })
    @Expose()
    attacks!: any;

    @Transform(({ obj }) => obj.system.classFeatLevels.value)
    @Expose()
    classFeatLevels!: number[];

    @Transform(({ obj }) => {
        return {
            unarmored: obj.system.defenses.unarmored,
            light: obj.system.defenses.light,
            medium: obj.system.defenses.medium,
            heavy: obj.system.defenses.heavy,
        };
    })
    @Expose()
    defenses!: any;

    @Transform(({ obj }) => {
        let description = obj.system.description.value;
        description = stepHeaders(description);
        description = assignHeaderIds(description);
        description = cleanupHTML(description);
        return description;
    })
    @Expose()
    description!: string;

    @Transform(({ obj }) => obj.system.generalFeatLevels.value)
    @Expose()
    generalFeatLevels!: number[];

    @Transform(({ obj }) => obj.system.hp)
    @Expose()
    hp!: number;

    @Transform(({ obj }) => obj.system.items)
    @Expose({ name: "items" })
    @Transform(({ obj }) => {
        type KeyValue = { key: string; value: any };
        const result: KeyValue[] = [];

        Object.entries(obj.system.items).forEach((value, _i) => {
            result.push({ key: value[0], value: value[1] } as KeyValue);
        });

        const items = result.map((v) => {
            return {
                refSlug: v.key,
                img: v.value.img,
                level: Number.parseInt(v.value.level),
                name: v.value.name,
                uuid: v.value.uuid,
            } as Item;
        }) as Item[];

        return items.map((item) => {
            const uuid = extractLastUuid(item.uuid);
            const feature = CLASS_FEATURES.get(uuid);
            if (feature) {
                return feature;
            }

            throw new Error(
                `assert failed: class feature '${item.name}' of class '${obj.system.slug}'  not found`,
            );
        });
    })
    @Type(() => Feat)
    features!: Feat[];

    @Transform(({ obj }) => obj.system.keyAbility.value)
    @Expose()
    keyAbility!: string[];

    @Transform(({ obj }) => obj.system.perception)
    @Expose()
    perception!: number;

    @Transform(({ obj }) => obj.system.publication)
    @Expose()
    publication!: Publication;

    @Transform(({ obj }) => obj.system.rules.map((v: any) => mapToRule(v)))
    @Expose()
    rules!: any[];

    @Transform(({ obj }) => obj.system.savingThrows)
    @Expose()
    savingThrows!: any;

    @Transform(({ obj }) => obj.system.skillFeatLevels.value)
    @Expose()
    skillFeatLevels!: number[];

    @Transform(({ obj }) => obj.system.skillIncreaseLevels.value)
    @Expose()
    skillIncreaseLevels!: number[];

    @Transform(({ obj }) => obj.system.spellcasting === 1)
    @Expose()
    spellcasting!: any;

    @Transform(({ obj }) => obj.system.trainedSkills)
    @Expose()
    trainedSkills!: any;

    @Transform(({ obj }) => obj.system.traits.rarity)
    @Expose()
    rarity!: string;

    @Transform(({ obj }) => obj.system.traits.value)
    @Expose()
    traits!: string[];

    @Transform(({ obj }) => obj.system.slug)
    @Expose()
    slug!: string;

    @Exclude() effects!: null;
    @Exclude() _stats!: null;

    @Exclude() img!: null;
    @Exclude() system!: null;
}
